File: software/examples/go/example_callback.go
package main

import (
	"fmt"
	"github.com/Tinkerforge/go-api-bindings/color_v2_bricklet"
	"github.com/Tinkerforge/go-api-bindings/ipconnection"
)

const ADDR string = "localhost:4223"
const UID string = "XYZ" // Change XYZ to the UID of your Color Bricklet 2.0.

func main() {
	ipcon := ipconnection.New()
	defer ipcon.Close()
	c, _ := color_v2_bricklet.New(UID, &ipcon) // Create device object.

	ipcon.Connect(ADDR) // Connect to brickd.
	defer ipcon.Disconnect()
	// Don't use device before ipcon is connected.

	c.RegisterColorCallback(func(r uint16, g uint16, b uint16, c uint16) {
		fmt.Printf("Color [R]: %d\n", r)
		fmt.Printf("Color [G]: %d\n", g)
		fmt.Printf("Color [B]: %d\n", b)
		fmt.Printf("Color [C]: %d\n", c)
		fmt.Println()
	})

	// Set period for color callback to 0.1s (100ms).
	c.SetColorCallbackConfiguration(100, false)

	fmt.Print("Press enter to exit.")
	fmt.Scanln()
}

File: software/examples/rust/example_callback.rs
use std::thread;
use std::{error::Error, io};
use tinkerforge::{color_v2_bricklet::*, ip_connection::IpConnection};

const HOST: &str = "localhost";
const PORT: u16 = 4223;
const UID: &str = "XYZ"; // Change XYZ to the UID of your Color Bricklet 2.0.

fn main() -> Result<(), Box<dyn Error>> {
    let ipcon = IpConnection::new(); // Create IP connection.
    let c = ColorV2Bricklet::new(UID, &ipcon); // Create device object.

    ipcon.connect((HOST, PORT)).recv()??; // Connect to brickd.
                                          // Don't use device before ipcon is connected.

    let color_receiver = c.get_color_callback_receiver();

    // Spawn thread to handle received callback messages.
    // This thread ends when the `c` object
    // is dropped, so there is no need for manual cleanup.
    thread::spawn(move || {
        for color in color_receiver {
            println!("Color [R]: {}", color.r);
            println!("Color [G]: {}", color.g);
            println!("Color [B]: {}", color.b);
            println!("Color [C]: {}", color.c);
            println!();
        }
    });

    // Set period for color callback to 0.1s (100ms).
    c.set_color_callback_configuration(100, false);

    println!("Press enter to exit.");
    let mut _input = String::new();
    io::stdin().read_line(&mut _input)?;
    ipcon.disconnect();
    Ok(())
}
